### DISCUSSION TITLE
Network displayed in visualization has correct nodes but wrong edges

### DISCUSSION THREAD

**Discussion by malawleyGH** [mesa/mesa#2134] [Q&A]
I am working on an ABM similar to the example virus on a network. The model itself is good, but the visualization is not working. The browser opens fine, a network is displayed, the nodes are correct, but the edges are all wrong. The portrayal method is creating the correct network configuration. I have traced the journey of the portrayal to the method "write_message" in websocket.py. My understanding is that this method converts the portrayal dictionary to json format and sends to the websocket. Here is a bit of the code:

def write_message(
        self, message: Union[bytes, str, Dict[str, Any]], binary: bool = False
    ) -> "Future[None]":
        """Sends the given message to the client of this Web Socket.

        The message may be either a string or a dict (which will be
        encoded as json).  If the ``binary`` argument is false, the
        message will be sent as utf8; in binary mode any byte string
        is allowed.

        If the connection is already closed, raises `WebSocketClosedError`.
        Returns a `.Future` which can be used for flow control.

        .. versionchanged:: 3.2
           `WebSocketClosedError` was added (previously a closed connection
           would raise an `AttributeError`)

        .. versionchanged:: 4.3
           Returns a `.Future` which can be used for flow control.

        .. versionchanged:: 5.0
           Consistently raises `WebSocketClosedError`. Previously could
           sometimes raise `.StreamClosedError`.
        """
        if self.ws_connection is None or self.ws_connection.is_closing():
            raise WebSocketClosedError()
        if isinstance(message, dict):
            message = tornado.escape.json_encode(message)
            print(f"Message 1 From websocket.py: {message}")
        return self.ws_connection.write_message(message, binary=binary)    

Notice that I inserted a print statement (near the bottom) that prints the message before it is sent. I checked the printed message carefully, it is the correct graph configuration, correct nodes and edges, correct node and edge colors, etc., and the json format of the message checks out when I tested it in python. 

When the network is displayed, the nodes are the correct color, but the edges are all wrong. I tried this in both Edge and Chrome and got the same result, so I don't think it is the browser. I am at a loss of what to do now. Any suggestions would be greatly appreciated, I will be happy to provide more information if needed.  Thanks

**Comment by rht** [comment]
This might be your problem? #1433. Otherwise, I recommend using the Solara-based new visualization, which will become the default in the upcoming Mesa release, 3.0.

**Reply by rht** [reply]
See https://github.com/projectmesa/mesa-examples/blob/main/examples/virus_on_network/app.py. To run it, `solara run app.py`.

**Reply by malawleyGH** [reply]
Thank you for this, but when I try to run it, I get the following:

PS C:\Users\malaw\Documents\MSDS\ship\model\virus_on_network\virus_on_network> solara run app.py
solara : The term 'solara' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path 
was included, verify that the path is correct and try again.
At line:1 char:1
+ solara run app.py
+ ~~~~~~
    + CategoryInfo          : ObjectNotFound: (solara:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

Solara is installed:

PS C:\Users\malaw\Documents\MSDS\ship\model\virus_on_network\virus_on_network> pip show solara
Name: solara
Version: 1.32.1
Summary:
Home-page:
Author:
Author-email: "Maarten A. Breddels" <maartenbreddels@gmail.com>
License: The MIT License (MIT)

        Copyright (c) 2022 Maarten A. Breddels

        Permission is hereby granted, free of charge, to any person obtaining a copy
        of this software and associated documentation files (the "Software"), to deal
        in the Software without restriction, including without limitation the rights
        to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
        copies of the Software, and to permit persons to whom the Software is
        furnished to do so, subject to the following conditions:

        The above copyright notice and this permission notice shall be included in
        all copies or substantial portions of the Software.

        THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
        IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
        FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
        AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
        LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
        OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
        THE SOFTWARE.
Location: c:\users\malaw\anaconda3\envs\sir\lib\site-packages
Requires: solara-server, solara-ui
Required-by: Mesa

I am running in a conda environment SIR. There is no executable for solara in the script directory: C:\Users\malaw\anaconda3\envs\SIR\Scripts

Here is the mesa version:

PS C:\Users\malaw\Documents\MSDS\ship\model\virus_on_network\virus_on_network> pip show mesa
Name: Mesa
Version: 2.3.0
Summary: Agent-based modeling (ABM) in Python
Home-page:
Author:
Author-email: Project Mesa Team <projectmesa@googlegroups.com>
License: Apache 2.0
Location: c:\users\malaw\anaconda3\envs\sir\lib\site-packages
Requires: click, cookiecutter, matplotlib, mesa-viz-tornado, networkx, numpy, pandas, solara, tqdm
Required-by:

**Reply by rht** [reply]
I don't have access to a Windows machine and so can't debug. @projectmesa/maintainers have you encountered this issue?

**Reply by malawleyGH** [reply]
I posted the above question on solara github and recieved this reply. 

__**"Hey @malawleyGH!

Did you upgrade from an older version of solara, or install in a fresh environment from scratch? We recently (see "Version 1.31.0" [here](https://solara.dev/changelog)) split the package to different subpackages. This could result in the solara command being removed, and would be fixed by reinstalling solara-server."_**_

This in fact is exactly what happened, I upgrade solara and the command "solara run app.py" would then not work.  

So, when I saw this, I uninstalled mesa and solara, then reinstalled. It seems I had to do this a couple of times. Then I started having other problems that I was able to resolve and eventually, the command "solara run app.py" worked and a browser opened, plotted the network, the sliders for input parameters, and a reactive chart. The "Reset", "Step", and "Reseed" buttons all seem to work.

But, the network itself is only drawn at the beginning. I need it to be redrawn on every step, since the nodes change state and thus nodes and edges change color. 

I think this is some of the appropriate code from jupyter_viz.py, with the very last else having the function drawing the network. Sorry, it will be hard to read since this editor removes all the indention.

def Card(model, measures, agent_portrayal, space_drawer, dependencies, color, layout_type):

    with rv.Card(style_=f"background-color: {color}; width: 100%; height: 100%") as main:
 
       if "Space" in layout_type:
            rv.CardTitle(children=["Space"])
            if space_drawer == "default":
                # draw with the default implementation
                components_matplotlib.SpaceMatplotlib(
                    model, agent_portrayal, dependencies=dependencies
                )

            elif space_drawer == "altair":
                components_altair.SpaceAltair(
                    model, agent_portrayal, dependencies=dependencies
                )

            elif space_drawer:
                # if specified, draw agent space with an alternate renderer
                space_drawer(model, agent_portrayal)

        elif "Measure" in layout_type:
            rv.CardTitle(children=["Measure"])
            measure = measures[layout_type["Measure"]]
           # The if and else below are under the elif above

            if callable(measure):
                # Is a custom object
                measure(model)

            else:
                components_matplotlib.PlotMatplotlib(
                    model, measure, dependencies=dependencies
                )
    return main

@solara.component
def JupyterViz(
    model_class,
    model_params,
    measures=None,
    name=None,
    agent_portrayal=None,
    space_drawer="default",
    play_interval=150,
    seed=None,):

    """Initialize a component to visualize a model.
    Args:
        model_class: class of the model to instantiate
        model_params: parameters for initializing the model
        measures: list of callables or data attributes to plot
        name: name for display
        agent_portrayal: options for rendering agents (dictionary)
        space_drawer: method to render the agent space for
            the model; default implementation is the `SpaceMatplotlib` component;
            simulations with no space to visualize should
            specify `space_drawer=False`
        play_interval: play interval (default: 150)
        seed: the random seed used to initialize the model
    """

My jupyterViz in app.py is below. I am just not seeing anything that would make me think the network visualization is being updated as the model steps. How would I the network to redraw on every step?  Thanks so very much.

page = JupyterViz(
    RegionModel,
    model_params,
    measures=[
        make_plot,
       make_text(get_active_ratio),
    ],
    name="Region Model",
    agent_portrayal=agent_portrayal,
 )

**Reply by rht** [reply]
There is not enough information from the code snippet you have provided. Have you tried the app.py in https://github.com/projectmesa/mesa-examples/tree/main/examples/virus_on_network verbatim?
